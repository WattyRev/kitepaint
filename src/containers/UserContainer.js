import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { getUser, getUserRecognition } from "../redux/modules/user";
import { SET_RECOGNIZED_USER, LOG_IN, LOG_OUT } from "../redux/actions";

/**
 * Maintains a local storage variable that indicates if the user is recognized as a registered user.
 * TODO this needs to be done in redux, because there will need to be more than one of these, and
 * they need to stay in sync.
 *
 * @param {Function} children
 */
class UserContainer extends React.Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
    onSetRecognition: PropTypes.func.isRequired,
    onLogIn: PropTypes.func.isRequired
  };

  /**
   * Toggles the variable in state and in local storage
   */
  toggleRecognition = () => {
    const { isRecognizedUser } = this.props;
    this.props.onSetRecognition(!isRecognizedUser);
  };

  render() {
    const {
      firstName,
      id,
      isLoggedIn,
      isLoggingIn,
      lastName,
      username
    } = this.props.user;
    return this.props.children({
      actions: {
        toggleRecognition: this.toggleRecognition,
        logIn: this.props.onLogIn,
        logOut: this.props.onLogOut
      },
      props: {
        firstName,
        id,
        isLoggedIn,
        isLoggingIn,
        isRecognizedUser: this.props.isRecognizedUser,
        lastName,
        username
      }
    });
  }
}

const mapStateToProps = state => ({
  user: getUser(state),
  isRecognizedUser: getUserRecognition(state)
});

const mapDispatchToProps = {
  onLogIn: LOG_IN,
  onLogOut: LOG_OUT,
  onSetRecognition: SET_RECOGNIZED_USER
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(UserContainer);
