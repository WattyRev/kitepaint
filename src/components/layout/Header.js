import React from "react";
import styled from "styled-components";
import UserContainer from "../../containers/UserContainer";
import { A, P, H2, Icon } from "../../theme";

export const StyleWrapper = styled.div`
  background: ${({ theme }) => theme.colors.silver};
  height: 48px;
  width: 100%;
  display: flex;
  align-items: center;
  padding: 0 8px;
  box-shadow: 0 0 1px ${({ theme }) => theme.colors.black};
  position: relative;
  box-sizing: border-box;

  .left,
  .right {
    display: flex;
    algin-items: center;
    flex-grow: 1;
  }
  .right {
    justify-content: flex-end;
  }
`;

/**
 * The header at the top of the application.
 */
const Header = () => (
  <StyleWrapper>
    <div className="left">
      <H2>
        <A href="#">KitePaint</A>
      </H2>
    </div>
    <div className="right">
      <UserContainer>
        {userData =>
          userData.props.isLoggedIn && (
            <P>
              <A onClick={userData.actions.logOut} className="testing_sign-out">
                Sign Out <Icon icon="sign-out-alt" />
              </A>
            </P>
          )
        }
      </UserContainer>
    </div>
  </StyleWrapper>
);

export default Header;
