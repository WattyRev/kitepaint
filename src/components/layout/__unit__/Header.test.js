import React from "react";
import { shallow, mount } from "enzyme";
import UserContainer from "../../../containers/UserContainer";
import Theme from "../../../theme";
import Header, { StyleWrapper } from "../Header";

describe("Header", () => {
  let defaultProps;
  beforeEach(() => {
    defaultProps = {};
  });
  it("renders", () => {
    shallow(<Header {...defaultProps} />);
  });
  it("should display the account dropdown if the user is logged in", () => {
    expect.assertions(1);
    const wrapper = shallow(<Header {...defaultProps} />);
    const userContainerContents = wrapper.find(UserContainer).prop("children")({
      actions: {
        logOut: jest.fn()
      },
      props: {
        isLoggedIn: true
      }
    });
    const userContainerWrapper = shallow(<div>{userContainerContents}</div>);
    expect(userContainerWrapper.find(".testing_account-dropdown")).toHaveLength(
      1
    );
  });
  it("should not display the account dropdown if the user is not logged in", () => {
    expect.assertions(1);
    const wrapper = shallow(<Header {...defaultProps} />);
    const userContainerContents = wrapper.find(UserContainer).prop("children")({
      actions: {
        logOut: jest.fn()
      },
      props: {
        isLoggedIn: false
      }
    });
    const userContainerWrapper = shallow(<div>{userContainerContents}</div>);
    expect(userContainerWrapper.find(".testing_account-dropdown")).toHaveLength(
      0
    );
  });
  describe("StyleWrapper", () => {
    it("renders", () => {
      mount(<StyleWrapper theme={Theme} />);
    });
  });
});
