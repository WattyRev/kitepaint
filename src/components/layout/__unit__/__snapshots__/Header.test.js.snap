// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders 1`] = `
<styled.div>
  <div
    className="left"
  >
    <Styled(H2)>
      <styled.a
        as={[Function]}
        to="/"
      >
        KitePaint
      </styled.a>
    </Styled(H2)>
  </div>
  <div
    className="right"
  >
    <styled.div
      as={[Function]}
      to="/"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon="home"
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        title=""
        transform={null}
      />
    </styled.div>
    <styled.div
      as={[Function]}
      to="/about"
    >
      About
    </styled.div>
    <styled.div
      as={[Function]}
      to="/create"
    >
      Create
    </styled.div>
    <Connect(UserContainer)>
      <Component />
    </Connect(UserContainer)>
  </div>
</styled.div>
`;
